### ISSUE TITLE
Error deploying to local iOS device

### ISSUE THREAD

**Issue by MattHoneycutt** [Microsoft/remotebuild#6]
I'm getting a "Cannot find the app installed on the device.  Check that the device is connected, awake, and unlocked" error when I try to debug my VS TACO app on an iPod attached to my Windows machine.  Here is the relevant info from the build log: 

2>------ Deploy started: Project: IntroToCordova2, Configuration: Debug iOS ------
2>Found iTunes version: 12.5.1.21
2>Uninstalling app ID:  io.cordova.introforpluralsight.
2>Successfully installed .ipa file:  E:\Projects\Pluralsight\IntroToCordova2\IntroToCordova2\bin\iOS\Debug\IntroToCordova2.ipa.
2>Starting idevicedebugserverproxy.exe at port 3000.
2>Cannot find the app installed on the device. Check that the device is connected, awake, and unlocked.
2>Deployment failed.

It claims it successfully installed, but I see no indication that it actually did.  

I've tried several different possible solutions online, and none have worked.  Any help?

**Comment by MSLaguana** [comment]
Is your iOS device running iOS 10? If so it may be because of that; we are working on updating the libimobiledevice binaries that we use to communicate with devices, but until we have that done we cannot debug iOS 10 devices.

**Comment by MattHoneycutt** [comment]
Thank you for the quick reply.  That must be it.  I am running iOS 10.0.2 on my mobile device.   

Any ETA on the update?   I'm working on a course covering VS TACO for Pluralsight, and I can delay recording the iOS modules if an update is coming soon-ish.  Otherwise, I'll just downgrade to iOS 9 on my device.  

Thank you again!

**Comment by jmatthiesen** [comment]
A fix is in the works and should be done "soon-ish," we'll keep you updated here!

**Comment by MattHoneycutt** [comment]
Awesome, thank you for the heads up!

**Comment by AviShmueli** [comment]
HI,
Is there something new with this issue?
I am facing the same problem...

**Comment by AlexanderDohmen** [comment]
Hi, we are facing the same trouble and would be very interested in a quick fix...

**Comment by an3urysm** [comment]
Hello, we are also blockedwith the same issue; our dev team works remotely and the team is unable to deploy to their local iOS 10.x devices after a successful build.

**Comment by MSLaguana** [comment]
We have been working on updating the `vs-libimobile` npm package with new binaries that support iOS 10. There were some unexpected difficulties, but we almost have a fix. Thanks for your patience.

**Comment by reinik** [comment]
We have the same probleme.

**Comment by reymonreyes** [comment]
I'm experiencing the same problem. I can't properly debug on VS. I hope an update is gonna be released soon.

**Comment by MattHoneycutt** [comment]
@MSLaguana Can we get an update on this?  Is there an ETA for when we can expect proper iOS 10/Xcode 8 support?

**Comment by kwiqniss** [comment]
I'm not posting in an official capacity at the moment, but have you tried in Visual Studio 2017 RC?

**Comment by MattHoneycutt** [comment]
No, I haven't, I'm one of the people affected by a bug in the VS 2017 installer that prevents it from installing the Cordova bits. :(

**Comment by MSLaguana** [comment]
@MattHoneycutt can you explain what you mean by that? Do you have a link to where you have reported/discussed any issues installing  VS2017?

**Comment by MSLaguana** [comment]
I also can't give any updates on when we can get out a fix other than Soon™, we want to get a fix out there asap.

**Comment by MattHoneycutt** [comment]
@MSLaguana Here's the bug report: https://developercommunity.visualstudio.com/content/problem/5646/vs-2017-rc-install-hangs-at-97-98-on-cordovawebtoo.html

Seems to be an issue with things kicked off via PowerShell.  I tried the workarounds (killing ssh-agent and anything else that I thought might be holding locks), but it didn't work for me.

**Comment by an3urysm** [comment]
agreed, we too are bleeding on this one...

**Comment by MSLaguana** [comment]
We have just published an update for this. If you are using Visual Studio 2017 RC, then the next time you start VS and attempt to launch for local iOS device, it should pick up the changes and work successfully.

If you are using Visual Studio 2015, then you will need to make some manual changes. Either go to Tools -> Options -> Tools for Apache Cordova and "Clear Cordova Cache", then perform a build to re-install the updated component (and everything else, it takes a little while), or you can go to `%appdata%\Roaming\npm\node_modules\vs-tac` and run `npm uninstall vs-libimobile && npm install vs-libimobile@0.0.4`

**Comment by MattHoneycutt** [comment]
@MSLaguana  Woohooo!  I'll give this a go later today.  Please pass along my thanks to the team!

**Comment by MattHoneycutt** [comment]
FYI, I'm still unable to debug an app connected to my local machine.  I tried both clearing the Cordova cache from the Visual Studio Options menu and manually updating to vs-libimobile@0.0.4 with npm.  Here's the output:

2>Found iTunes version: 12.5.4.42
2>Uninstalling app ID:  com.heroicapplications.vstaco.sample.
2>Successfully installed .ipa file:  E:\Projects\Pluralsight\IntroToCordova\IntroToCordova\IntroToCordova\bin\iOS\Debug\PS Intro.ipa.
2>Starting idevicedebugserverproxy.exe at port 3011.
2>Timed out connecting to local iOS device.
2>Failed to launch the app on the device. Check that the device is connected, awake, and unlocked.
2>Starting ios_webkit_debug_proxy.exe at port 9221.
2>Deployment failed.

The app does indeed install on my connected device, but it never launches.  

I verified that vs-libimobile 0.0.4 is installed in the node_modules folder of ```%appdata%\Roaming\npm\node_modules\vs-tac%```.

Is there any other info I can provide?  Are any updates needed on the OS X machine running remotebuild?

**Comment by reymonreyes** [comment]
I also tried both and now I'm getting a lot of errors.
1>------ Build started: Project: SitePro, Configuration: Debug Android ------
1>  ------ Ensuring correct global installation of package from source package directory: C:\Program Files (x86)\Microsoft Visual Studio 14.0\Common7\IDE\Extensions\ApacheCordovaTools\packages\vs-tac
1>  ------ Name from source package.json: vs-tac
1>  ------ Version from source package.json: 1.0.42
1>  ------ Package not currently installed globally.
1>  ------ Installing globally from source package. This could take a few minutes...
1>  Each package is licensed to you by its owner. Microsoft is not responsible for, nor does it grant any licenses to, third-party packages. Some packages may include dependencies which are governed by additional licenses. Follow the package source (feed) URL to determine any dependencies.
1>  npm ERR! Windows_NT 10.0.14393
1>  npm ERR! argv "C:\\Users\\Admin\\AppData\\Roaming\\Microsoft\\VisualStudio\\MDA\\vs-npm\\2.14.9\\\\node.exe" "C:\\Users\\Admin\\AppData\\Roaming\\Microsoft\\VisualStudio\\MDA\\vs-npm\\2.14.9\\node_modules\\npm\\bin\\npm-cli.js" "-g" "install" "C:\\Program Files (x86)\\Microsoft Visual Studio 14.0\\Common7\\IDE\\Extensions\\ApacheCordovaTools\\packages\\vs-tac" "--loglevel" "warn"
1>  npm ERR! node v0.12.9
1>  npm ERR! npm  v2.14.9
1>  npm ERR! path C:\Users\Admin\AppData\Roaming\npm\node_modules\vs-tac
1>  npm ERR! code EBUSY
1>  npm ERR! errno -4082
1>  npm ERR! EBUSY, rmdir 'C:\Users\Admin\AppData\Roaming\npm\node_modules\vs-tac'
1>  npm ERR!
1>RUNMDAINSTALL : npm ERR! If you need help, you may report this error at:
1>  npm ERR!     <https://github.com/npm/npm/issues>
1>RUNMDAINSTALL : npm ERR! error rolling back error : EBUSY, rmdir 'C:\Users\Admin\AppData\Roaming\npm\node_modules\vs-tac'
1>  npm ERR! error rolling back     at Error (native)
1>  npm ERR! error rolling back  { [Error: EBUSY, rmdir 'C:\Users\Admin\AppData\Roaming\npm\node_modules\vs-tac']
1>  npm ERR! error rolling back   errno: -4082,
1>  npm ERR! error rolling back   code: 'EBUSY',
1>  npm ERR! error rolling back   path: 'C:\\Users\\Admin\\AppData\\Roaming\\npm\\node_modules\\vs-tac' }
1>  npm ERR! Please include the following file with any support request:
1>  npm ERR!     C:\Users\Admin\AppData\Roaming\Microsoft\VisualStudio\MDA\vs-npm\2.14.9\npm-debug.log
1>  ------ npm install failed. Exit code: 4294963214
1>  ------ Package installation failed. Retrying...
1>  npm ERR! Windows_NT 10.0.14393
1>  npm ERR! argv "C:\\Users\\Admin\\AppData\\Roaming\\Microsoft\\VisualStudio\\MDA\\vs-npm\\2.14.9\\\\node.exe" "C:\\Users\\Admin\\AppData\\Roaming\\Microsoft\\VisualStudio\\MDA\\vs-npm\\2.14.9\\node_modules\\npm\\bin\\npm-cli.js" "-g" "uninstall" "vs-tac" "--loglevel" "warn"
1>  npm ERR! node v0.12.9
1>  npm ERR! npm  v2.14.9
1>  npm ERR! path C:\Users\Admin\AppData\Roaming\npm\node_modules\vs-tac
1>  npm ERR! code EBUSY
1>  npm ERR! errno -4082
1>  npm ERR! EBUSY, rmdir 'C:\Users\Admin\AppData\Roaming\npm\node_modules\vs-tac'
1>  npm ERR!
1>RUNMDAINSTALL : npm ERR! If you need help, you may report this error at:
1>  npm ERR!     <https://github.com/npm/npm/issues>
1>  npm ERR! Please include the following file with any support request:
1>  npm ERR!     C:\Users\Admin\AppData\Roaming\Microsoft\VisualStudio\MDA\vs-npm\2.14.9\npm-debug.log
1>  ------ npm uninstall failed. Exit code: 4294963214
1>  npm ERR! Windows_NT 10.0.14393
1>  npm ERR! argv "C:\\Users\\Admin\\AppData\\Roaming\\Microsoft\\VisualStudio\\MDA\\vs-npm\\2.14.9\\\\node.exe" "C:\\Users\\Admin\\AppData\\Roaming\\Microsoft\\VisualStudio\\MDA\\vs-npm\\2.14.9\\node_modules\\npm\\bin\\npm-cli.js" "-g" "install" "C:\\Program Files (x86)\\Microsoft Visual Studio 14.0\\Common7\\IDE\\Extensions\\ApacheCordovaTools\\packages\\vs-tac" "--loglevel" "warn"
1>  npm ERR! node v0.12.9
1>  npm ERR! npm  v2.14.9
1>  npm ERR! path C:\Users\Admin\AppData\Roaming\npm\node_modules\vs-tac
1>  npm ERR! code EBUSY
1>  npm ERR! errno -4082
1>  npm ERR! EBUSY, rmdir 'C:\Users\Admin\AppData\Roaming\npm\node_modules\vs-tac'
1>  npm ERR!
1>RUNMDAINSTALL : npm ERR! If you need help, you may report this error at:
1>  npm ERR!     <https://github.com/npm/npm/issues>
1>RUNMDAINSTALL : npm ERR! error rolling back error : EBUSY, rmdir 'C:\Users\Admin\AppData\Roaming\npm\node_modules\vs-tac'
1>  npm ERR! error rolling back     at Error (native)
1>  npm ERR! error rolling back  { [Error: EBUSY, rmdir 'C:\Users\Admin\AppData\Roaming\npm\node_modules\vs-tac']
1>  npm ERR! error rolling back   errno: -4082,
1>  npm ERR! error rolling back   code: 'EBUSY',
1>  npm ERR! error rolling back   path: 'C:\\Users\\Admin\\AppData\\Roaming\\npm\\node_modules\\vs-tac' }
1>  npm ERR! Please include the following file with any support request:
1>  npm ERR!     C:\Users\Admin\AppData\Roaming\Microsoft\VisualStudio\MDA\vs-npm\2.14.9\npm-debug.log
1>  ------ npm install failed. Exit code: 4294963214
1>  ------ Package installation failed. Retrying...
1>  npm ERR! Windows_NT 10.0.14393
1>  npm ERR! argv "C:\\Users\\Admin\\AppData\\Roaming\\Microsoft\\VisualStudio\\MDA\\vs-npm\\2.14.9\\\\node.exe" "C:\\Users\\Admin\\AppData\\Roaming\\Microsoft\\VisualStudio\\MDA\\vs-npm\\2.14.9\\node_modules\\npm\\bin\\npm-cli.js" "-g" "uninstall" "vs-tac" "--loglevel" "warn"
1>  npm ERR! node v0.12.9
1>  npm ERR! npm  v2.14.9
1>  npm ERR! path C:\Users\Admin\AppData\Roaming\npm\node_modules\vs-tac
1>  npm ERR! code EBUSY
1>  npm ERR! errno -4082
1>  npm ERR! EBUSY, rmdir 'C:\Users\Admin\AppData\Roaming\npm\node_modules\vs-tac'
1>  npm ERR!
1>RUNMDAINSTALL : npm ERR! If you need help, you may report this error at:
1>  npm ERR!     <https://github.com/npm/npm/issues>
1>  npm ERR! Please include the following file with any support request:
1>  npm ERR!     C:\Users\Admin\AppData\Roaming\Microsoft\VisualStudio\MDA\vs-npm\2.14.9\npm-debug.log
1>  ------ npm uninstall failed. Exit code: 4294963214
1>C:\Program Files (x86)\MSBuild\Microsoft\VisualStudio\v14.0\ApacheCordovaTools\vs-mda-targets\Microsoft.MDA.FileMirroring.targets(285,5): error : Error installing local npm package.
1>MSBUILD : cordova-build error : module.js:338
1>  module.js:338
1>MSBUILD : cordova-build error :  throw err;
1>      throw err;
1>MSBUILD : cordova-build error :  ^
1>            ^
1>MSBUILD : cordova-build error BLD401: Error : BLD00401 : Could not find module 'C:\Users\Admin\AppData\Roaming\npm\node_modules\vs-tac\app.js'. Please Go to Tools --> Options --> Tools for Apache Cordova --> Cordova Tools --> Clear Cordova Cache and try building again.
1>  Error Cannot find module 'C:\Users\Admin\AppData\Roaming\npm\node_modules\vs-tac\app.js'
1>MSBUILD : cordova-build error :  at Function.Module._resolveFilename (module.js:336:15)
1>      at Function.Module._resolveFilename (module.js:336:15)
1>MSBUILD : cordova-build error :  at Function.Module._load (module.js:278:25)
1>      at Function.Module._load (module.js:278:25)
1>MSBUILD : cordova-build error :  at Function.Module.runMain (module.js:501:10)
1>      at Function.Module.runMain (module.js:501:10)
1>MSBUILD : cordova-build error :  at startup (node.js:129:16)
1>      at startup (node.js:129:16)
1>MSBUILD : cordova-build error :  at node.js:814:3
1>      at node.js:814:3

**Comment by MattHoneycutt** [comment]
@reymonreyes I got that error, too, initially.  Make sure you don't have any instances of iosdebugproxy.exe running.  If you see them in task manager (or anything like that), go ahead and kill them.  That did the trick for me.

**Comment by reymonreyes** [comment]
@MattHoneycutt Thanks for the tip. I also did a rebuild and everything went okay. But, I'm still getting an error "Timed out launching the app. Check your network connection to the mac and try again. If you are launching to a device, make sure you have Web Inspector enabled on it." It's still not working.

**Comment by MSLaguana** [comment]
What version of iOS are you running on your device, and do you have a corresponding entry in your device support folder?

**Comment by MSLaguana** [comment]
If you are targeting iOS 10.2, then you'll need to have the device support libraries from Xcode 8.2 to be able to launch apps on a device.

If you can't get a hold of those libraries, there is a workaround for you though: Hit f5 to build once, have the deploy fail after installing the app on your device. Then manually start the app, and go to `Build -> Configuration Manager` in visual studio and uncheck deploy (and build if you want). Then hit f5 and it should attach to the app running on the device. Make sure to check those boxes again afterwards so that it builds and deploys the next time you need to.

**Comment by reymonreyes** [comment]
@MSLaguana The last one you suggested did work. It worked on my device first and then it after that it also started working correctly on the simulator.

**Comment by MattHoneycutt** [comment]
@MSLaguana Sorry for the late reply, I'm still catching up from the holidays!

I was indeed missing the 10.2 device support folder.  Once I acquired that, everything worked as expected.  Thanks!

**Comment by mdjones3** [comment]
I had struggled with this same issue but am now able to launch a debug session on my local IOS device after the fix from MSLaguana.  Thanks!  The only problem is that the debug session won't stop on any break points.  Is anyone else seeing this and is there something I need to do to fix this?
Thanks

**Comment by MattHoneycutt** [comment]
@mdjones3 Did you try running window.location.reload() in the VS JavaScript console?  I don't think breakpoints will be hit on the first load for iOS applications.

**Comment by MSLaguana** [comment]
Breakpoints should be hit on the first load, but only if the event that you are breaking on happens after we connect. There is no way to start a Cordova application "paused", so as soon as the application starts the javascript code begins running. Since there is a slight delay before we can attach to the app, this means that some events that happen early in the application's lifecycle happen before we are debugging, so they get missed and you need to do something like the `window.location.reload()` mentioned above. However if you set a breakpoint on an event that happens later on, such as a button press event, or you `setTimeout(function () { console.log("breakpoint here";}, 10000);` or something then the breakpoints should be hit.

Does this account for what you see, @mdjones3 ?

**Comment by mdjones3** [comment]
No, unfortunately I'm unable to break on the app at any point.  This is only a problem running a local IOS device.  I have no such problem with a local Android device or in Windows.

If this isn't a universal problem, I can poke around with my local setup.  FYI I'm running VS2015 update 3.

**Comment by MSLaguana** [comment]
Could you please open a new issue so we don't have any confusion about exactly what the problem is, including details on what iOS you are running and whether this happens with a trivial app? It'd also be great if you could try using the debugger to modify the running app, such as tweaking `document.body.style.background = "red"` or something like that and see how that behaves. I'll see if I can replicate the issue.

**Comment by mdjones3** [comment]
Yes, I will open a new issue.  @MattHoneycutt, I tried calling window.location.reload() which does reload the app but I still can't break.  @MSLaguana I am able to set the background to red from the Javascript console so it appears I can modify the running app.  I'll add these comments to the issue.

Thanks for the quick responses!

**Comment by reymonreyes** [comment]
I've also been experiencing this one. My breakpoints aren't hitting on iOS. I'm running iOS 10.2.

**Comment by ekambarrao** [comment]
@MSLaguana  I have been fighting to build and debug since last 30 days.

as per this link

https://stackoverflow.com/questions/43130683/vs2017-cordova-ios-local-device-deploy

there should be "vs-libimobile" in node_modules folder but I am not able to find any of that.

Do I have to use any plugin specific to ios for building and debugging?
